### PR TITLE
feat: add numberOfChunks and Chunked on record

### LINKED ISSUE #310: Chunked and NumberOfChunked are not saved in the Datastore
**What happened**:
Chunked and NumberOfChunks are missing from the records that were created when initializing a chunk upload session,

**What you expected to happen**:
When passing Chunked and NumberofChunks to Open Saves, I expect the properties to be saved against the record in the table for "record" entities.

**How to reproduce it (as minimally and precisely as possible)**:
Initialize a chunk upload with all required properties, including Chunked and NumberOfChunks.
When getting the record (regardless if chunks are uploaded or not), the properties are missing from the record.

**Anything else we need to know?**:

**Environment**:
- Open Saves version: latest
- Kubernetes version (use `kubectl version`): N/A
- Cloud provider or hardware configuration: GCP
- Install method (yaml/helm): yaml

### PR DESCRIPTION
**Pull request by vasconcelosvcd**
/kind feat

**What this PR does / Why we need it**:
This PR adds `Chunked` and `numberOfChunks` on record model to be saved against datastore 

**Which issue(s) this PR fixes**:
Closes #310

### CODE REVIEW

**Review by loranger2k — APPROVED** [review]
LGTM.

File: internal/pkg/metadb/record/record_test.go [hunk 1]
```diff
@@ -265,8 +280,10 @@ func TestRecord_NewRecordFromProto(t *testing.T) {
 	createdAt := time.Date(1992, 1, 15, 3, 15, 55, 0, time.UTC)
 	updatedAt := time.Date(1992, 11, 27, 1, 3, 11, 0, time.UTC)
 	proto := &pb.Record{
-		Key:      "key",
-		BlobSize: int64(len(testBlob)),
+		Key:            "key",
+		BlobSize:       int64(len(testBlob)),
+		Chunked:        false,
+		NumberOfChunks: 1,
```
Comment (yuryu) [inline internal/pkg/metadb/record/record_test.go:286]: I would not populate NumberOfChunks if it's not chunked (the object does not have a chunk). How about using 0 as a placeholder? I think that's the behavior of the server code as well.

File: internal/pkg/metadb/record/record_test.go [hunk 2]
```diff
@@ -284,9 +301,11 @@ func TestRecord_NewRecordFromProto(t *testing.T) {
...
 	}
 	expected := &Record{
-		Key:          "key",
-		BlobSize:     int64(len(testBlob)),
-		ExternalBlob: uuid.Nil,
+		Key:            "key",
+		BlobSize:       int64(len(testBlob)),
+		ExternalBlob:   uuid.Nil,
+		Chunked:        false,
+		NumberOfChunks: 1,
```
Comment (yuryu) [inline internal/pkg/metadb/record/record_test.go:308]: ditto

File: internal/pkg/metadb/record/record.go [hunk 3]
```diff
@@ -30,14 +30,16 @@ var _ datastore.PropertyLoadSaver = new(PropertyMap)
...
-	ExternalBlob uuid.UUID `datastore:"-"`
-	Properties   PropertyMap
-	OwnerID      string
-	Tags         []string
-	OpaqueString string `datastore:",noindex"`
+	Key            string `datastore:"-"`
+	Blob           []byte `datastore:",noindex"`
+	BlobSize       int64
+	ExternalBlob   uuid.UUID `datastore:"-"`
+	NumberOfChunks int64
```
Comment (yuryu) [inline internal/pkg/metadb/record/record.go:37]: Do we want int64 for this? I think int32 is reasonably big.

**Review by yuryu — COMMENTED** [review]
A couple of comments but overall LGTM. Do we have code to actually populate these fields? Is it coming later?

**Comment by yuryu** [comment]
On second thought, I think these fields should be populated by the server, not something passed from the client (just like #305). What do you think?

**Review by yuryu — CHANGES_REQUESTED** [review]
Let me mark this as request changes.

File: internal/pkg/metadb/record/record.go (hunk 3 above)
Comment (loranger2k) [reply internal/pkg/metadb/record/record.go:37]: Yes, int32 is more than enough.

File: internal/pkg/metadb/record/record_test.go (hunk 1 above)
Comment (loranger2k) [reply internal/pkg/metadb/record/record_test.go:286]: Good observation.

File: internal/pkg/metadb/record/record.go (hunk 3 above)
Comment (yuryu) [reply internal/pkg/metadb/record/record.go:37]: Okay I looked at the proto and it looks like it's int64.

Comment (vasconcelosvcd) [reply internal/pkg/metadb/record/record.go:37]: I've kept int64 due the need to convert it to 32/64 back an forward, but if the conversions are ok, sure its big enough

**Comment by yuryu** [comment]
If we are making these two fields read-only, we change the comments in the proto:
https://github.com/googleforgames/open-saves/blob/99041ce2f34490bc5e03adea3c2bae43685663c3/api/open_saves.proto#L216

And update Record.Chunked/NumberOfChunks here
https://github.com/googleforgames/open-saves/blob/99041ce2f34490bc5e03adea3c2bae43685663c3/internal/pkg/metadb/metadb.go#L496

The bigger problem is we don't have tests for chunked uploads (#282) so we've got to do this too (can be separate but preferably soon).

**Comment by loranger2k** [comment]
Are you suggesting to pass the Chunked and NumberOfChunks when calling CreateChunkedBlobRequest?

How is the clients gong to know when initializing the download how many chunks to expect if it is not saved against the record?

**Comment by loranger2k** [comment]
Alternatively, we could add NumberOfChunks to the CreateChunkedBlobRequest and save it against the Record (along with Chunked=true). We would always ignore both Chunked and NumberOfChunks in the create/update record, but expose it in the GetRecord response.

**Comment by yuryu** [comment]
No, these variables should be read-only for the client. The server will count the number of chunks when committing (just like it populates the `Size` field) and store the value to appropriate entities. The client should not need to send `Chunked`, `NumberOfChunks`, `Size` etc when creating a record or uploading a blob.

Because we use the same gRPC message types, we need to have these fields in `Record` anyways if we want to return these values with `GetRecord`.

The overall flow would look like this:

- client: `CreateRecord()`
  - server: ignores `chunked` and `number_of_chunks`
- client: `CreateChunkedBlob()`, ..., `CommitChunkedUpload()`
  - server: calculates the values for `number_of_chunks` and `chunked` and stores them to `Record` datastore entity.
- later, client: `GetRecord()`
  - server: reads the values from the datastore `Record` entity and returns them to the client.
 
What do you think?

**Comment by loranger2k** [comment]
That works. @yuryu Would you be able to help us with this implementation?

**Comment by loranger2k** [comment]
If so, we can simply close this PR and open another one when changes are ready.

**Comment by loranger2k** [comment]
Also, I propose we include chunk validation with hash values (when present) before incrementing the NumberOfChunks on the server.

**Comment by yuryu** [comment]
Sure, I'll first add tests so that we can validate the messages.

File: internal/pkg/metadb/record/record.go [hunk 4]
```diff
@@ -100,14 +102,16 @@ func (r *Record) LoadKey(k *datastore.Key) error {
...
-		Tags:         r.Tags,
-		Properties:   r.Properties.ToProto(),
-		OpaqueString: r.OpaqueString,
-		CreatedAt:    timestamppb.New(r.Timestamps.CreatedAt),
-		UpdatedAt:    timestamppb.New(r.Timestamps.UpdatedAt),
+		Key:            r.Key,
+		BlobSize:       r.BlobSize,
+		OwnerId:        r.OwnerID,
+		Chunked:        r.Chunked,
+		NumberOfChunks: r.NumberOfChunks,
```
Comment (yuryu) [inline internal/pkg/metadb/record/record.go:109]: We will need to remove these two lines in the final implementation

**Pull request closed without merging**